src/main/java/frc/robot/subsystems/RelaySubsystem.java

```diff
@@ -0,0 +1,198 @@
+import edu.wpi.first.wpilibj.Relay;
+import edu.wpi.first.wpilibj.Relay.Value;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.SubsystemChecker;
+import frc.robot.SubsystemChecker.SubsystemType;
+import frc.robot.config.Config;
+
+/**
+ * Vision Subsystem to handle any hardware on robot side that is needed
+ * for vision. Such as:
+ * 
+ * - Relay for ring light
+ * 
+ */
+public class RelaySubsystem extends SubsystemBase {
+
+    // static variable single_instance for a Singleton class 
+    private static RelaySubsystem single_instance = null;
+
+    /** 
+     * Get Instance of vision subsystem 
+     */
+    public static RelaySubsystem getInstance() {
+        if (single_instance == null) 
+            SubsystemChecker.subsystemConstructed(SubsystemType.RelaySubsystem);
+            single_instance = new RelaySubsystem(); 
+  
+        return single_instance; 
+    }
+
+    private final ArrayList<NTRelay> m_relays;
+
+    /**
+     * Construct the instance of this class
+     */
+    private RelaySubsystem() {
+        /**
+         * Set up hardware stuff
+         */
+        m_relays = new ArrayList<NTRelay>();
+        
+        // Setup relays to be controlled from network tables
+        addNTRelay(Config.RELAY_RINGLIGHT_REAR_SMALL, "Ringlight_Rear_Small");
+        addNTRelay(Config.RELAY_RINGLIGHT_REAR_LARGE, "Ringlight_Rear_Large");
+        addNTRelay(Config.RELAY_RINGLIGHT_FRONT, "Ringlight_Front");
+
+    }
+
+    private class NTRelay {
+        private int m_portNumber;
+        private Relay m_relay;
+        private BooleanEntry m_entryNT;
+        boolean isOn;
+
+        public NTRelay(int portNumber, String topicName) {
+            m_portNumber = portNumber;
+            m_relay = new Relay(portNumber);
+            
+            BooleanTopic topic = NetworkTableInstance.getDefault()
+                .getTable(Config.RELAY_NETWORKTABLE)
+                .getBooleanTopic(topicName);
+            topic.setPersistent(true);
+
+            
+            m_entryNT = topic.getEntry(isOn);
+
+            isOn = m_entryNT.get(false);
+            setRelay(isOn);
+        }
+
+        public void checkNT() {
+            boolean entryValue = m_entryNT.get();
+            if (entryValue != isOn) {
+                setRelay(entryValue);
+            }
+        }
+
+        public void setRelay(boolean turnOn) {
+            isOn = turnOn;
+            if (isOn) {
+                m_relay.set(Value.kOn);
+            } else {
+                m_relay.set(Value.kOff);
+            }
+            updateNT();
+        }
+
+        public void toggleRelay() {
+            isOn = !isOn;
+            setRelay(isOn);
+        }
+
+        private void updateNT() {
+            m_entryNT.accept(isOn);
+        }
+
+        public int getPortNumber() {
+            return m_portNumber;
+        }
+    }
+
+    @Override
+    public void periodic() {
+        for (int i = 0; i < m_relays.size(); i++) {
```

Comment:
This part is using RoboRio resource. 
During competition, we don't need to use Network table entry to turn on/off relays. 
We could use odometry to control ring lights on/off or just turn them on all the time, it they don't impact AprilTag camera.

Reply:
I actually disagree on this one. It's just monitoring 3 networktable topics which is really just checking for a boolean value.

This will be needed if we want vision to be able to control the ringlights. We will need to communicate with them about who controls the ringlights. It's very likely that robot code will choose when to use vision and can also turn on the ringlight as the same time but we need to have these conversations.